cpp/include/cuspatial/detail/algorithm/linestring_distance.cuh

```diff
@@ -83,17 +97,62 @@ __global__ void linestring_distance_multilinestring_loop(MultiLinestringRange1 m
+                                                  SegmentRange2 multilinestrings2,
+                                                  IndexRange thread_bounds,
                                                   thrust::optional<uint8_t*> intersects,
-                                                  OutputIt distances_first)
+                                                  OutputIt distances)
 {
-  using T = typename MultiLinestringRange1::element_t;
+  using index_t = typename IndexRange::value_type;
 
-  for (auto idx = threadIdx.x + blockIdx.x * blockDim.x; idx < multilinestrings1.num_points();
+  auto num_segment_pairs = thread_bounds[thread_bounds.size() - 1];
+  for (auto idx = threadIdx.x + blockIdx.x * blockDim.x; idx < num_segment_pairs;
        idx += gridDim.x * blockDim.x) {
-    atomicMin(&distances_first[geometry_idx], static_cast<T>(sqrt(min_distance_squared)));
+
+    auto [it, geometry_id] = compute_geometry_id(thread_bounds.begin(), thread_bounds.end(), idx);
+
+    auto first_thread_of_this_block = blockDim.x * blockIdx.x;
+    auto last_thread_of_block = blockDim.x * (blockIdx.x + 1) - 1;
+    auto first_thread_of_next_geometry = thread_bounds[geometry_id + 1];
+
+    bool split_block = first_thread_of_this_block < first_thread_of_next_geometry && first_thread_of_next_geometry <= last_thread_of_block;
+
+    if (intersects.has_value() && intersects.value()[geometry_id]) {
+      distances[geometry_id] = 0.0f;
+      continue;
+    }
+
+    auto const local_idx = idx - *it;
+    // Retrieve the number of segments in multilinestrings[geometry_id]
+    auto num_segment_this_multilinestring =
+      multilinestrings1.multigeometry_count_begin()[geometry_id];
+    // The segment id from multilinestring1 this thread is computing (local_id + global_offset)
+    auto multilinestring1_segment_id = local_idx % num_segment_this_multilinestring +
+                                       multilinestrings1.multigeometry_offset_begin()[geometry_id];
+
+    // The segment id from multilinestring2 this thread is computing (local_id + global_offset)
+    auto multilinestring2_segment_id = local_idx / num_segment_this_multilinestring +
+                                       multilinestrings2.multigeometry_offset_begin()[geometry_id];
+
+    auto [a, b] = multilinestrings1.begin()[multilinestring1_segment_id];
+    auto [c, d] = multilinestrings2.begin()[multilinestring2_segment_id];
+
+    auto partial = sqrt(squared_segment_distance(a, b, c, d));
+
+    if (split_block)
+      atomicMin(&distances[geometry_id], partial);
+    else
+    {
+      // block reduce
+      typedef cub::BlockReduce<T, block_size> BlockReduce;
+      __shared__ typename BlockReduce::TempStorage temp_storage;
+      auto aggregate = BlockReduce(temp_storage).Reduce(partial, cub::Min());
+
+      // atmomic with leading thread
+      if (cooperative_groups::this_thread_block().thread_rank() == 0)
```

Comment:
Is there any advantage to this over `if (threadIdx.x == 0)`?  What code does this produce? It's impossible for threadIdx.x == 0 to have exited before any other threads in its block in this code.

Reply:
Cannot think of any advantage comparing to `threadIdx.x == 0`. I originally had hopped writing modularized kernel with cg but but later regressed. This is probably a remnant from prototyping.